authentication/oidc/oidc.go

```diff
@@ -0,0 +1,126 @@
+/*
+ * Copyright 2018 The Service Manager Authors
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package oidc
+
+import (
+	"context"
+	"encoding/json"
+	"errors"
+	"fmt"
+	"io/ioutil"
+	"net/http"
+	"net/url"
+	"strings"
+
+	"github.com/Peripli/service-manager/authentication"
+	"github.com/coreos/go-oidc"
+	"github.com/sirupsen/logrus"
+)
+
+// Options is the configuration used to construct a new OIDC authenticator
+type Options struct {
+	// IssuerURL is the base URL of the token issuer
+	IssuerURL string
+
+	// ClientID is the id of the oauth client used to verify the tokens
+	ClientID string
+
+	// ReadConfigurationFunc is the function used to call the token issuer. If one is not provided, http.DefaultClient.Do will be used
+	ReadConfigurationFunc DoRequestFunc
+}
+
+// DoRequestFunc is an alias for any function that takes an http request and returns a response and error
+type DoRequestFunc func(request *http.Request) (*http.Response, error)
+
+// Authenticator is the OpenID implementation of authentication.Authenticator
+type Authenticator struct {
+	Verifier authentication.TokenVerifier
+}
+
+// NewAuthenticator returns a new OpenID authenticator or an error if one couldn't be configured
+func NewAuthenticator(ctx context.Context, options Options) (authentication.Authenticator, error) {
+	// Work around for UAA until https://github.com/cloudfoundry/uaa/issues/805 is fixed
+	// Then oidc.NewProvider(ctx, options.IssuerURL) should be used
+	if _, err := url.ParseRequestURI(options.IssuerURL); err != nil {
+		return nil, err
+	}
+	wellKnown := strings.TrimSuffix(options.IssuerURL, "/") + "/.well-known/openid-configuration"
+	req, err := http.NewRequest(http.MethodGet, wellKnown, nil)
+	if err != nil {
+		return nil, err
+	}
+
+	var readConfigFunc DoRequestFunc
+	if options.ReadConfigurationFunc != nil {
+		readConfigFunc = options.ReadConfigurationFunc
+	} else {
+		readConfigFunc = http.DefaultClient.Do
```

Comment:
Default client does not have request timeouts set. We should go against the option to use it by default

Reply:
I think it's nice to have defaults. However, I strongly agree the default shouldn't be used in production.

Reply:
Actually, if a request with context that has timeouts set is provided it will be sufficient